src/js/interpreter_cache.h

```diff
@@ -0,0 +1,83 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "./interpreter_cache_interface.h"
+#include "ccf/pal/locking.h"
+#include "ds/lru.h"
+
+namespace ccf::js
+{
+  class InterpreterCache : public AbstractInterpreterCache
+  {
+  protected:
+    // Locks access to all internal fields
+    ccf::pal::Mutex lock;
+    LRU<std::string, std::shared_ptr<js::Context>> lru;
+    size_t cache_build_marker;
+
+  public:
+    InterpreterCache(size_t max_cache_size) : lru(max_cache_size) {}
+
+    std::shared_ptr<js::Context> get_interpreter(
+      js::TxAccess access,
+      const JSDynamicEndpoint& endpoint,
+      size_t freshness_marker) override
+    {
+      if (access != js::TxAccess::APP)
+      {
+        throw std::logic_error(
+          "JS interpreter reuse lru is only supported for APP "
+          "interpreters");
+      }
+
+      std::lock_guard<ccf::pal::Mutex> guard(lock);
+
+      if (cache_build_marker != freshness_marker)
+      {
+        LOG_INFO_FMT(
+          "Clearing interpreter lru at {} - rebuilding at {}",
+          cache_build_marker,
+          freshness_marker);
+        lru.clear();
+        cache_build_marker = freshness_marker;
+      }
+
+      if (endpoint.properties.interpreter_reuse.has_value())
+      {
+        switch (endpoint.properties.interpreter_reuse->kind)
+        {
+          case ccf::endpoints::InterpreterReusePolicy::KeyBased:
+          {
+            const auto key = endpoint.properties.interpreter_reuse->key;
+            auto it = lru.find(key);
+            if (it == lru.end())
+            {
+              LOG_TRACE_FMT(
+                "Inserting new interpreter into cache, with key {}", key);
+              it = lru.insert(key, std::make_shared<js::Context>(access));
+            }
+            else
+            {
+              LOG_TRACE_FMT(
+                "Returning interpreter previously in cache, with key {}", key);
+              lru.promote(it);
+            }
+
+            return it->second;
+          }
+        }
```

Comment:
Do we need a `default:` here or is that for forward compatibility?

Reply:
kind is already an enum at this point, the validation will have taken place in the JSON deserialisation. So this is not forward compatible, but we also don't particularly need a default.